### PR TITLE
cli: add ability to specify a name instead of primary key

### PR DESCRIPTION
**Pull request by ryanpetrello**
I'm of the opinion that - while not perfect in its handling of uniqueness rules - this is a behavior we should _carry over_ from tower-cli, because it makes life **much** easier on the user.

### CODE REVIEW

**Comment by ryanpetrello** [comment]
image

image

image

File: awxkit/awxkit/cli/__init__.py [hunk 1]
```diff
@@ -62,6 +62,9 @@ def run(stdout=sys.stdout, stderr=sys.stderr, argv=[]):
                     allow_unicode=True
                 )
             ))
+        elif cli.get_config('format') == 'human':
```
Comment (ryanpetrello) [inline awxkit/awxkit/cli/__init__.py:65]: Just an omission I noticed while working on this.

**Review by wenottingham — APPROVED** [review]
feature approved. did not review code.

**Comment by wenottingham** [comment]
I assume this does not allow using a named url fragment as a name?

**Comment by AlanCoding** [comment]
If you want an example of confusion from accepting names versus named URL "slugs", see https://github.com/ansible/tower-cli/issues/735. I believe it would be a very bad idea to take named URLs for this.

Question on the function - if I have a job template named "42" with a pk value of 56, I would assume that there's no way for me to reference it by name?

Another question, let's say that I have an object, and I want to `awx modify` a related foreign key so that it is null. How do I do that? Imagine that I wanted to make a JT prompt for inventory, then set its inventory to be null.

Right now, I can't manage to do that (trying 0, null, None, etc.)

```
awx job_template modify 95 --inventory 0
{"inventory": ["Invalid pk \"0\" - object does not exist."]}
```

Since this gets into the business of automatic detection of the type of the content fed into the parameter, it's a good time to answer this question. Does type checking gauntlet turn into "is it an integer?", "is it null-ish?", "could it be a named URL?", "is it maybe a name?"

I would prefer that a related lookup by name could be implemented as a totally different option from primary keys. To be direct, I also expect to be overruled on that.

In lieu of named URL slugs, I would like to develop a primitive proposal for using multiple fields for the lookup, thus respecting the functionality of named URLs, but in a format clearly exposed to the CLI user. The basic parameters of that would likely be...

 - OPTIONS returns a new field that gives the uniqueness rules of the resource (same as named URL rules)
 - args and related options accept `nargs='*'`
 - depending on the number of args given, it goes from the most local field to the most remote field. For example, first argument is name, second argument is organization name.

That proposal is not inconsistent with the implementation here (would be a superset), so would not block merging this. But it can offer a roadmap to justify why to not do named URLs now.

File: awxkit/awxkit/cli/options.py [hunk 2]
```diff
@@ -1,23 +1,72 @@
...
+        return int(value)
+
+    identity = 'name'
+
+    if not page:
+        if not hasattr(v2, model_name):
+            if model_name in DEPRECATED_RESOURCES_REVERSE:
+                model_name = DEPRECATED_RESOURCES_REVERSE[model_name]
+
+        if model_name == 'users':
```
Comment (jakemcdermott) [inline awxkit/awxkit/cli/options.py]: :+1:

File: awxkit/awxkit/cli/options.py [hunk 3]
```diff
@@ -1,23 +1,72 @@
...
+            identity = 'username'
+        elif model_name == 'instances':
+            model_name = 'hostname'
```
Comment (jakemcdermott) [inline awxkit/awxkit/cli/options.py:32]: :+1: hostnames, usernames  are the main naming corner cases I'm aware of besides workflow nodes and credential input sources (which have no names, only pks)

**Review by jakemcdermott — APPROVED** [review]

**Comment by jakemcdermott** [comment]
nameless resources seem to be handled as well:
```shell
~/projects/awx$ awx --conf.username admin --conf.password password --conf.host 'https://127.0.0.1:8043' -k workflow_job_template_nodes get "foo"
{"detail": "WorkflowJobTemplateNode has no field named 'name'"}
```

**Comment by ryanpetrello** [comment]
@AlanCoding @wenottingham 

All good points and worth considering going forward.  But I do want to stress that given the time available, my goal here is feature parity with `tower-cli`, including some of the warts you've mentioned (especially regarding @AlanCoding's thoughts on improvements).

**Pull request merged**